### PR TITLE
[Bugfix] Fix DeepSeek V4 mHC broadcast buffer for dummy load

### PR DESCRIPTION
**Pull request by HollowMan6**
## Purpose

With `--load-format dummy`, `load_weights()` is skipped, so
`hc_attn_fn_broadcast` stays `None` and the first decoder layer's 2-D
broadcast path fails at `assert self.hc_attn_fn_broadcast is not None`.

Fix: define a model-level `process_weights_after_loading()` on
`DeepseekV4ForCausalLM`. The loader invokes this hook for every load
format — including dummy load — so the broadcast buffer is always
populated before the first forward.

The hook also finalizes MegaMoE weights, replacing
the lazy `finalize_weights()` call in the MegaMoE forward that existed only
to cover dummy load. This takes the branch out of the forward hot path and
moves the heavy weight repack (which also frees the original params) out of
the first forward, so profiling sees steady-state memory. The EPLB-side
guard call in `get_expert_mapping` is unchanged.

Because the MTP and DSpark drafters load as their own top-level models
(registry → `get_model` → loader → PWAL), each gets a matching hook —
the removed forward-time call was their only finalize path under dummy
load.

The companion fix for weight refit (in-place `copy_()` refresh so the
buffer address stays stable for captured CUDA graphs) is split into a
separate PR.

## Purpose

With `--load-format dummy`, `load_weights()` is skipped, so
`hc_attn_fn_broadcast` stays `None` and the first decoder layer's 2-D
broadcast path fails at `assert self.hc_attn_fn_broadcast is not None`.

Fix: define a model-level `process_weights_after_loading()` on
`DeepseekV4ForCausalLM`. The loader invokes this hook for every load
format — including dummy load — so the broadcast buffer is always
populated before the first forward.

Per review (@aoshen02), the hook also finalizes MegaMoE weights, replacing
the lazy `finalize_weights()` call in the MegaMoE forward that existed only
to cover dummy load. This takes the branch out of the forward hot path and
moves the heavy weight repack (which also frees the original params) out of
the first forward, so profiling sees steady-state memory. The EPLB-side
guard call in `get_expert_mapping` is unchanged.

Because the MTP and DSpark drafters load as their own top-level models
(registry → `get_model` → loader → PWAL), each gets a matching hook —
the removed forward-time call was their only finalize path under dummy
load.

The companion fix for weight refit (in-place `copy_()` refresh so the
buffer address stays stable for captured CUDA graphs) is split into a
separate PR.

## Not a duplicate

Searched open PRs for this area: #50645 guards the broadcast kernel on
DeepGEMM support, #49707 fixes mHC warmup coverage (consumes the buffer,
doesn't populate it), #47807 streamlines warmup on the kernel side. None
touch load-time weight finalization or MegaMoE finalize placement.

## Test Plan

Two unit tests added to `tests/models/test_deepseek_v4_mega_moe.py`:
- `test_deepseek_v4_pwal_hook_finalizes_mega_moe_and_mhc_broadcast` — the
  target model's hook runs both finalizes.
- `test_deepseek_v4_drafter_pwal_hooks_finalize_mega_moe` — the MTP and
  DSpark hooks delegate to their MegaMoE finalize.

End to end tests on verl side with DSV4 training

## Test Result

- New tests: **2 passed**; full file: **9 passed** (GPU).
- Mutation check: with the three model files reverted to base, both new
  tests fail — they pin the hooks, not incidentals.
- `ruff check` + `ruff format --check` clean.

End to end tests on verl side with DSV4 training returned to normal training inference mismatch.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [X] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [X] The test plan, such as providing test command.
- [X] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/models/deepseek_v4/nvidia/model.py [hunk 1]
```diff
@@ -1373,11 +1376,15 @@ def finalize_mhc_broadcast_weights(self) -> None:
...
-            layer.hc_attn_fn_broadcast = (
+            broadcast = (
                 layer.hc_attn_fn.detach()
                 .view(-1, layer.hc_mult, layer.hidden_size)
                 .sum(dim=1)
             )
+            if layer.hc_attn_fn_broadcast is None:
+                layer.hc_attn_fn_broadcast = broadcast
+            else:
+                layer.hc_attn_fn_broadcast.copy_(broadcast)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply vllm/models/deepseek_v4/nvidia/model.py:1379]: This is over defense.

File: vllm/models/deepseek_v4/nvidia/model.py [hunk 2]
```diff
@@ -1105,6 +1105,9 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
         else:
             self._mtp_hidden_buffer = None
 
+        # Populate hc_attn_fn_broadcast for dummy load (load_weights is skipped).
+        self.finalize_mhc_broadcast_weights()
```
Comment (aoshen02) [inline vllm/models/deepseek_v4/nvidia/model.py]: Maybe we should put it in PWAL, not only this model file has such a problem.

Comment (HollowMan6) [reply vllm/models/deepseek_v4/nvidia/model.py]: Thanks for reviewing, now I moved this call to `process_weights_after_loading`. For now `hc_attn_fn_broadcast` exists only in the nvidia DeepSeek V4 file, other model files don't have the problem.

File: vllm/models/deepseek_v4/nvidia/model.py [hunk 3]
```diff
@@ -1544,5 +1548,9 @@ def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
         self.model.finalize_mhc_broadcast_weights()
         return loaded_params
 
+    def process_weights_after_loading(self) -> None:
```
Comment (aoshen02) [inline vllm/models/deepseek_v4/nvidia/model.py:1545]: Will self.model.finalize_mhc_broadcast_weights been called the second time when do pwal during normal weight loading?

File: vllm/models/deepseek_v4/nvidia/model.py [hunk 4]
```diff
@@ -1544,5 +1548,9 @@ def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
...
+        # Populate hc_attn_fn_broadcast for dummy load.
```
Comment (aoshen02) [inline vllm/models/deepseek_v4/nvidia/model.py]: Do you think it's better to do:
def process_weights_after_loading(self) -> None:
    self.model.finalize_mega_moe_weights()
    self.model.finalize_mhc_broadcast_weights()

and remove self.finalize_weights() inside the megamoe forward

File: vllm/models/deepseek_v4/nvidia/model.py (hunk 3 above)
Comment (HollowMan6) [reply vllm/models/deepseek_v4/nvidia/model.py:1545]: yes, on normal weight loading, `finalize_mhc_broadcast_weights()` runs twice, but the second run changes nothing as when #52626 is merged, so it should be Ok.

File: vllm/models/deepseek_v4/nvidia/model.py (hunk 4 above)
Comment (HollowMan6) [reply vllm/models/deepseek_v4/nvidia/model.py]: Yes, those can be cleaner, but under dummy load, the forward-time finalize_weights() is the only call that finalizes the MTP/DSpark drafters, as they have no PWAL hooks of their own. Removing it breaks dummy-load spec-dec unless we add hooks to both drafters too, which will need more testing and I think they should be done in a separate refactoring PR, so I would rather keep this PR scoped to the mHC broadcast buffer only.

Comment (HollowMan6) [reply vllm/models/deepseek_v4/nvidia/model.py]: Done — the hook now finalizes both MegaMoE and mHC broadcast weights, and the lazy finalize_weights() in the MegaMoE forward is removed. I also added the same hook to the MTP and DSpark drafters: they load as their own top-level models, so the forward-time call was their only finalize path under dummy load. Unit tests added for all three hooks.

File: vllm/models/deepseek_v4/nvidia/model.py (hunk 3 above)
Comment (HollowMan6) [reply vllm/models/deepseek_v4/nvidia/model.py:1545]: ~now removed in load_weights~ this will break verl.

**Comment by aoshen02** [comment]
LGTM, @WoosukKwon could you take a look?
@HollowMan6 Please add a gsm8k test result before and after this PR and make sure that init with dummy weight loader and then reload, the gsm8k is the same as normal weight loader.

**Comment by HollowMan6** [comment]
@aoshen02 Now I verified this with verl e2e and found that removing finalize* from load_weights https://github.com/vllm-project/vllm/pull/51368/commits/177101df1ff3783259c86e3e854817f01c5029b0 will actually break the sync, so I reverted the change in the last commit, and it should still be fine: https://github.com/vllm-project/vllm/pull/51368#discussion_r3797361602.

After the fix, I checked with verl e2e again by init with dummy weight loader and then reload with the normal weight load, and the training inference mismatch (log probs diff) are consistent and reasonably low at step 1, so I think we are good to go now:

```bash
rollout_actor_probs_pearson_corr:0.9952459931373596 rollout_probs_diff_mean:0.0063540199771523476
```

**Review by WoosukKwon — APPROVED** [review]
